interfaces/apparmor/spec.go

```diff
@@ -88,6 +90,92 @@ func RegisteredSnippetKeys() []string {
 	return keylist
 }
 
+// MetadataTag is an opaque string used to tag a snippet.
+//
+// Metadata tags can be added to AppArmor rules or snippets to encode additional
+// information which the kernel can then send back to snapd when it sends a
+// message notification.
+//
+// When each metadata tag is registered, it must either be associated with a
+// particular interface or associated with no interface.
+//
+// Once a tag is registered in association with an interface, any attempt to
+// register it in association with a different interface or with no interface
+// at all will result in a panic, though a tag may be re-registered in
+// association with the same interface with which it was originally registered.
+//
+// Likewise, once registered as unassociated with any interface, attempts to
+// re-register the same tag in association with an interface will result in a
+// panic, but the tag may be re-registered as unassociated again.
+type MetadataTag struct {
+	tag string
+}
+
+func (mt *MetadataTag) String() string {
+	return mt.tag
+}
+
+var validMetadataTagRegexp = regexp.MustCompile("^[a-zA-Z0-9/_-]+$")
+
+func newMetadataTag(tag string) MetadataTag {
+	if !validMetadataTagRegexp.MatchString(tag) {
+		logger.Panicf("cannot register invalid metadata tag: %q", tag)
+	}
+	return MetadataTag{tag: tag}
+}
+
+// registeredMetadataTags is a mapping from metadata tag to the interface with
+// which it was associated when registered. If the tag was registered as
+// unassociated with any interface, its value in the map is "".
+var registeredMetadataTags map[MetadataTag]string = make(map[MetadataTag]string)
+
+const tagUnassociated = ""
+
+// RegisterMetadataTagWithInterface marks the given tag as being associated
+// with the given interface and returns the tag as a MetadataTag. If the tag
+// is already associated with a different interface, or it is explicitly
+// registered as unassociated with any interface, panics with an error.
+func RegisterMetadataTagWithInterface(tag string, iface string) MetadataTag {
+	if iface == "" {
+		logger.Panicf("cannot register metadata tag %q to empty string; use RegisterMetadataTagWithoutInterface instead", tag)
+	}
+	metadataTag := newMetadataTag(tag)
+	if val, ok := registeredMetadataTags[metadataTag]; ok && val != iface {
+		if val == tagUnassociated {
```

Comment:
using a const here makes me think there's a special non-zero value in use
```suggestion
		if val == "" {
```